packetbeat/sniffer/sniffer.go

```diff
@@ -168,16 +308,42 @@ func (s *Sniffer) Run() error {
 
 		data, ci, err := handle.ReadPacketData()
 		if err == pcap.NextErrorTimeoutExpired || isAfpacketErrTimeout(err) { //nolint:errorlint // pcap.NextErrorTimeoutExpired is not wrapped.
-			logp.Debug("sniffer", "timedout")
+			logp.Debug("sniffer", "timed out")
+
+			// If we have timed out too many times and we are following
+			// a default route, request a new default route interface.
+			const maxTimeouts = 10 // Place-holder until we have a sensible notion of how big this should be.
+			timeouts++
+			if s.followDefault && timeouts > maxTimeouts {
+				select {
+				case refresh <- struct{}{}:
+				default:
+					// Don't request to refresh if already requested.
+				}
+				timeouts = 0
+			}
 			continue
 		}
+		timeouts = 0
 
 		if err != nil {
 			// ignore EOF, if sniffer was driven from file
 			if err == io.EOF && s.config.File != "" { //nolint:errorlint // io.EOF should never be wrapped.
 				return nil
 			}
 
+			// If we are following a default route, request an interface
+			// refresh and log the error.
+			if s.followDefault {
+				select {
+				case refresh <- struct{}{}:
+				default:
+					// Don't request to refresh if already requested.
+				}
+				logp.Warn("error during packet capture: %v", err)
+				continue
+			}
+
 			s.state.Store(snifferInactive)
 			return fmt.Errorf("sniffing error: %w", err)
```

Comment:
This raises the question of how we treat errors on non-default route interfaces. I think it's arguable that they should also just keep trying.

Reply:
🤔  I can make arguments for it both ways.

If the interfaces goes way temporarily for some reason then I still want Packetbeat to resume when it comes back. I could let systemd/launchd/agent handle restarting the process continuously.

On the other hand, it is useful to have hard failure when testing out my config. I can quickly catch that I typo'ed the interface name if Packetbeat exits immediately.

Reply:
OK, so an option would be to keep a state of whether there was ever a connection. If there was, keep trying, but if there wasn't, fail out. I would not want to add this in this change, but it is an option to consider in the future.